### PR TITLE
Rendering Pipeline Improvements

### LINKED ISSUE #167: end_shape() is closing shapes without 'CLOSE'
Consider/run this code:

```
from p5 import *

window_w = 1500
window_h = 1000
center_x = window_w/2
center_y = window_h/2


def delta_coords(degrees, distance):
	angle_radians = degrees * math.pi / 180
	x = math.cos(angle_radians) * distance
	y = math.sin(angle_radians) * distance
	return x, y

def setup():
	size(window_w, window_h)
	background(255)


def draw():
	background(255)
	for n in range(1):
		start_angle = 0
		end_angle = 500
		start_radius = 350

		curr_raidus = start_radius

		no_fill()
		begin_shape()
		for d in range(start_angle, end_angle):

			final_distance = curr_raidus
			d_x, d_y = delta_coords(d, final_distance)

			x = center_x + d_x
			y = center_y + d_y

			vertex(x, y)

			curr_raidus += .02

		end_shape()

	no_loop()

run()



```
This creates a very gentle spiral. It appears that once the end angle > 360, the shape closes. This should not happen unless end_shape('CLOSE') is specified. 

**System information:**
 - p5 release - 0.6.0
 - Python version - 3.7
 - Operating system - OSX

### LINKED ISSUE #188: Inconsistent behavior for Arc
**Describe the bug**
The behavior for Arc in p5py does match with the Processing Language

**To Reproduce**
```python
from p5 import *

def draw():
    offset = 0
    x = 0
    while x < 2 * PI:
        arc((50.0 + offset, 50.0), 50.0, 50.0, x, 2 * PI)
        arc((50.0 + offset, 100), 50, 50, x, 2 * PI, 'OPEN')
        arc((50.0 + offset, 150), 50, 50, x, 2 * PI, 'CHORD')
        arc((50.0 + offset, 200), 50, 50, x, 2 * PI, 'PIE')
        x += QUARTER_PI
        offset += 50

if __name__ == '__main__':
    run()
```

```java
void setup() {
  size(640, 360);
}

void draw() {
  int offset = 0;
  for (float x = 0; x < PI * 2; x += QUARTER_PI) {
     arc(50.0 + offset, 50.0, 50.0, 50.0, x, 2*PI);
     arc(50.0 + offset, 100, 50, 50, x, 2*PI, OPEN);
     arc(50.0 + offset, 150, 50, 50, x, 2*PI, CHORD);
     arc(50.0 + offset, 200, 50, 50, x, 2*PI, PIE);
     offset += 50;
  }
}
```

**Expected behavior**
Screen Shot 2020-06-22 at 11 47 09 PM

**Actual Behavior**
Screen Shot 2020-06-22 at 11 47 25 PM

**System information:**
 - p5 release (version number or latest commit): 975f07e75a48aa7ebc8225a9cb5b27fda8795bb8
 - Python version: 3.8.3
 - Operating system: MacOS 10.15.5

### PR DESCRIPTION
**Pull request by ziyaointl**
Apologies in advance for sending in a huge PR like this. I found it difficult to divide the GLUTess change into smaller PRs without breaking a lot of the current functionality. Huge thanks to anyone who's willing to read through the changes, even just parts of it. Suggestions and comments are really appreciated.

### TLDR
This PR achieves the following:
1. Make the rendering pipeline a bit simpler to understand
2. Use PyOpenGL for tessellation & remove the triangle library
3. Fixes #188 
4. Fixes #167 
5. Add a profile harness that enables easy profiling via cProfile & a way to quantify performance
6. Add test scenes and records performance improvements

### Details
1. Make the rendering pipeline a bit simpler to understand
Here's a diagram of the new pipeline that I'm proposing.
p5py pipeline
The crucial part of this new pipeline is update_draw queue. It takes a PShape instance and generates meshes and lines which can then be fed to their rendering functions. It also invokes GLUTess if needed.

2. Use PyOpenGL for tessellation & remove the triangle library
This means adding a Tessellator class and removing `_retriangulate` and related instance variables in `PShape`.

3. Fixes #188
arcs

4. Fixes #167 
Screen Shot 2020-06-28 at 12 11 56 AM

5. Profile harness & performance improvements
The harness works by importing a list of scenes one-by-one, modifying their draw functions to exit after rendering 100 frames, and evoking cProfile while p5.run() is running. I still need to understand why it crashes on certain sequences of scenes, but when it works the results seems to be consistent.

Here are the timings for rendering 100 frames of each scene. The scenes can be found in the profiling folder.

| Scene          | Old    | New   | Delta   |
|----------------|--------|-------|---------|
| arcs           | 8.82s  | 8.79s |  -0.34% |
| triangle_strip | 20.77s | 1.36s | -93.45% |
| custom_shapes  | 2.86s  | 2.79s |  -2.45% |
| custom_shapes2 | 3.85s  | 1.87s | -51.43% |
| curves         | 2.19s  | 2.74s |  25.11% |
| primitives     | 1.58s  | 1.69s |   6.96% |

### Known Issues & Regressions
(Will address in later PRs)
1. `update_draw_queue` should be implemented in the renderer, not in `PShape`
2. The 3D renderer can't render 2D shapes now
3. `shape.edit` needs to be implemented
4. Need to check for the minimum number of vertices is provided in `update_draw_queue`

### CODE REVIEW

File: profiling/arcs.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    x = 0
+    while x < 2 * PI:
+        arc((50.0 + offset, 50.0), 50.0, 50.0, x, 2 * PI)
+        arc((50.0 + offset, 100), 50, 50, x, 2 * PI, 'OPEN')
+        arc((50.0 + offset, 150), 50, 50, x, 2 * PI, 'CHORD')
+        arc((50.0 + offset, 200), 50, 50, x, 2 * PI, 'PIE')
+        x += QUARTER_PI
+        offset += 50
+
+    # no_fill()
```
Comment (parsoyaarihant) [inline profiling/arcs.py:17]: remove comments?

**Comment by parsoyaarihant** [comment]
I am getting the following error

```
  File "/Users/arihantparsoya/Documents/p5/p5/core/tess.py", line 1, in <module>
    from OpenGL.GLU import gluNewTess, gluTessCallback, GLU_TESS_VERTEX, GLU_TESS_BEGIN, GLU_TESS_END, GLU_TESS_ERROR, \
ModuleNotFoundError: No module named 'OpenGL'
```

Can you add `OpenGL` as a dependency?

File: profiling/triangle_strip.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        global x, y, outsideRadius, insideRadius
+        background(204)
+
+        numPoints = 60
+        angle = 0
+        angleStep = 180.0 / numPoints
+
+        begin_shape("TRIANGLE_STRIP")
+
+        for i in range(numPoints):
```
Comment (parsoyaarihant) [inline profiling/triangle_strip.py:26]: I think it should be `range(numPoints + 1)` as the strip is not being closed:

Screenshot 2020-06-30 at 6 45 46 AM

**Comment by parsoyaarihant** [comment]
The code looks good. Can you fix the errors on PyFlakes (for the files you worked on)?

```python
➜  p5 git:(tessellation) ✗ pyflakes p5/
p5/__init__.py:19: 'from .sketch import *' used; unable to detect undefined names
p5/__init__.py:19: '.sketch.*' imported but unused
p5/__init__.py:20: 'from .core import *' used; unable to detect undefined names
p5/__init__.py:20: '.core.*' imported but unused
p5/__init__.py:21: 'from .pmath import *' used; unable to detect undefined names
p5/__init__.py:21: '.pmath.*' imported but unused
p5/__init__.py:23: '.__version__.__title__' imported but unused
p5/__init__.py:24: '.__version__.__description__' imported but unused
p5/__init__.py:25: '.__version__.__url__' imported but unused
p5/__init__.py:26: '.__version__.__version__' imported but unused
p5/__init__.py:27: '.__version__.__author__' imported but unused
p5/__init__.py:28: '.__version__.__license__' imported but unused
p5/__init__.py:29: '.__version__.__copyright__' imported but unused
p5/__main__.py:19: '._p5' imported but unused
p5/core/primitives3d.py:19: 'functools' imported but unused
p5/core/primitives3d.py:22: '.p5' imported but unused
p5/core/primitives3d.py:26: '.primitives.draw_shape' imported but unused
p5/core/primitives.py:595: local variable 'valid_values' is assigned to but never used
p5/core/__init__.py:18: 'from .structure import *' used; unable to detect undefined names
p5/core/__init__.py:18: '.structure.*' imported but unused
p5/core/__init__.py:19: 'from .shape import *' used; unable to detect undefined names
p5/core/__init__.py:19: '.shape.*' imported but unused
p5/core/__init__.py:20: 'from .primitives import *' used; unable to detect undefined names
p5/core/__init__.py:20: '.primitives.*' imported but unused
p5/core/__init__.py:21: 'from .primitives3d import *' used; unable to detect undefined names
p5/core/__init__.py:21: '.primitives3d.*' imported but unused
p5/core/__init__.py:22: 'from .transforms import *' used; unable to detect undefined names
p5/core/__init__.py:22: '.transforms.*' imported but unused
p5/core/__init__.py:23: 'from .color import *' used; unable to detect undefined names
p5/core/__init__.py:23: '.color.*' imported but unused
p5/core/__init__.py:24: 'from .attribs import *' used; unable to detect undefined names
p5/core/__init__.py:24: '.attribs.*' imported but unused
p5/core/__init__.py:25: 'from .image import *' used; unable to detect undefined names
p5/core/__init__.py:25: '.image.*' imported but unused
p5/core/__init__.py:26: 'from .font import *' used; unable to detect undefined names
p5/core/__init__.py:26: '.font.*' imported but unused
p5/core/__init__.py:27: 'from .constants import *' used; unable to detect undefined names
p5/core/__init__.py:27: '.constants.*' imported but unused
p5/core/__init__.py:28: 'from .vertex import *' used; unable to detect undefined names
p5/core/__init__.py:28: '.vertex.*' imported but unused
p5/core/__init__.py:29: 'from .svg import *' used; unable to detect undefined names
p5/core/__init__.py:29: '.svg.*' imported but unused
p5/core/vertex.py:19: 'numpy as np' imported but unused
p5/core/vertex.py:24: 'from .constants import *' used; unable to detect undefined names
p5/core/vertex.py:52: 'SType' may be undefined, or defined from star imports: .constants
p5/core/vertex.py:53: 'SType' may be undefined, or defined from star imports: .constants
p5/core/vertex.py:55: 'SType' may be undefined, or defined from star imports: .constants
p5/core/vertex.py:302: 'SType' may be undefined, or defined from star imports: .constants
p5/core/vertex.py:320: 'SType' may be undefined, or defined from star imports: .constants
p5/core/vertex.py:325: 'SType' may be undefined, or defined from star imports: .constants
p5/core/vertex.py:330: 'SType' may be undefined, or defined from star imports: .constants
p5/core/svg.py:92: local variable 'stroke_cap' is assigned to but never used
p5/core/svg.py:127: local variable 'stroke_cap' is assigned to but never used
p5/core/svg.py:162: local variable 'width' is assigned to but never used
p5/core/svg.py:163: local variable 'height' is assigned to but never used
p5/core/attribs.py:27: 'from .constants import *' used; unable to detect undefined names
p5/core/attribs.py:27: '.constants.*' imported but unused
p5/core/shape.py:25: 'triangle as tr' imported but unused
p5/core/shape.py:28: 'from .constants import *' used; unable to detect undefined names
p5/core/shape.py:105: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:475: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:479: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:479: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:479: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:479: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:484: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:489: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:501: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:506: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:510: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:514: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:518: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:522: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:526: 'SType' may be undefined, or defined from star imports: .constants
p5/core/shape.py:528: 'SType' may be undefined, or defined from star imports: .constants
p5/core/image.py:22: 'time' imported but unused
p5/core/image.py:278: local variable 'te' is assigned to but never used
p5/core/tests/test_svg.py:5: 'p5.core.svg.parse_circle' imported but unused
p5/core/tests/test_svg.py:5: 'p5.core.svg.parse_ellipse' imported but unused
p5/core/tests/test_svg.py:5: 'p5.core.svg.parser' imported but unused
p5/core/tests/test_svg.py:7: 'p5.core.shape.PShape' imported but unused
p5/pmath/__init__.py:19: 'from .utils import *' used; unable to detect undefined names
p5/pmath/__init__.py:19: '.utils.*' imported but unused
p5/pmath/__init__.py:20: 'from .vector import *' used; unable to detect undefined names
p5/pmath/__init__.py:20: '.vector.*' imported but unused
p5/pmath/__init__.py:21: 'from .rand import *' used; unable to detect undefined names
p5/pmath/__init__.py:21: '.rand.*' imported but unused
p5/pmath/__init__.py:22: 'from .curves import *' used; unable to detect undefined names
p5/pmath/__init__.py:22: '.curves.*' imported but unused
p5/pmath/__init__.py:23: 'from .time import *' used; unable to detect undefined names
p5/pmath/__init__.py:23: '.time.*' imported but unused
p5/pmath/__init__.py:24: '.matrix' imported but unused
p5/pmath/curves.py:23: 'math' imported but unused
p5/sketch/userspace.py:36: 'from ..core.constants import *' used; unable to detect undefined names
p5/sketch/userspace.py:36: '..core.constants.*' imported but unused
p5/sketch/__init__.py:21: 'from .userspace import *' used; unable to detect undefined names
p5/sketch/__init__.py:21: '.userspace.*' imported but unused
p5/sketch/renderer2d.py:34: 'from ..core.constants import *' used; unable to detect undefined names
p5/sketch/renderer2d.py:34: '..core.constants.*' imported but unused
p5/sketch/renderer3d.py:35: 'from ..core.constants import *' used; unable to detect undefined names
p5/sketch/renderer3d.py:309: 'Z_EPSILON' may be undefined, or defined from star imports: ..core.constants
p5/sketch/base.py:172: use ==/!= to compare str, bytes, and int literals
```
